src/distributions/range2.rs

```diff
@@ -181,37 +182,116 @@ range_int_impl! { usize, usize }
+    // than positive numbers
+    NegAndPos { cutoff: X, scale: X },
 }
 
 macro_rules! range_float_impl {
-    ($ty:ty) => {
+    ($ty:ty, $next_u:path) => {
         impl SampleRange for $ty {
             type T = RangeFloat<$ty>;
         }
-        
+
         impl RangeImpl for RangeFloat<$ty> {
+            // We can distinguish between two different kinds of ranges:
+            //
+            // Completely positive or negative.
+            // A characteristic of these ranges is that they get more bits of
+            // precision as they get closer to 0.0. For positive ranges it is as
+            // simple as applying a scale and offset to get the right range.
+            // For a negative range, we apply a negative scale to transform the
+            // range [0,1) to (-x,0]. Because this results in a range with it
+            // closed and open sides reversed, we do not sample from
+            // `closed_open01` but from `open_closed01`.
+            //
+            // A range that is both negative and positive.
+            // This range has as characteristic that it does not have most of
+            // its bits of precision near its low or high end, but in the middle
+            // near 0.0. As the basis we use the [-1,1) range from
+            // `closed_open11`.
+            // How it works is easiest to explain with an example.
+            // Suppose we want to sample from the range [-20, 100). We are
+            // first going to scale the range from [-1,1) to (-60,60]. Note the
+            // range changes from closed-open to open-closed because the scale
+            // is negative.
+            // If the sample happens to fall in the part (-60,-20), move it to
+            // (-100,60). Then multiply by -1.
+            // Scaling keeps the bits of precision intact. Moving the assymetric
+            // part also does not waste precision, as the more you move away
+            // from zero the less precision can be stored.
+
             type X = $ty;
-            
+
             fn new(low: Self::X, high: Self::X) -> Self {
-                RangeFloat {
-                    low: low,
-                    range: high - low,
+                if low >= 0.0 {
+                    RangeFloat::Positive {
+                        offset: low,
+                        scale: high - low,
+                    }
+                } else if high <= 0.0 {
+                    RangeFloat::Negative {
+                        offset: high,
+                        scale: low - high,
+                    }
+                } else if -low <= high {
+                    RangeFloat::PosAndNeg {
+                        cutoff: low,
+                        scale: (high + low) / -2.0 + low,
+                    }
+                } else {
+                    RangeFloat::NegAndPos {
+                        cutoff: high,
+                        scale: (high + low) / 2.0 - high,
```

Comment:
It seems the half-open `[low, high)` range works well here, doesn't it? Is a closed version, `[low, high]` feasible? Where `low` is close to zero, the same approximation trick, `(low, high]` may be okay (and equiv for high < 0), but for ranges spanning 0 there's no good solution?

The other question is whether this actually matters much in practice... many applications don't need so much precision in the first place, but maybe some do?

Reply:
I have spend some hours trying to get the rounding at the edge of the ranges right, but without much success yet... So all the stuff about open/closed is theoretically nice, but not yet the reality :-(. That is one of the reasons I was not sure in the other pr if it made sense to expose a closed range in the treat, because it does not work really out for floats.

But I will give my test for de edge cases a few more tries. You asked above, if I am done tweaking. Maybe not yet completely ;-)

Reply:
This is also a good rationale for not worrying too much about which open/closed distributions we have. I checked the "pure math" version of these distributions gives the right open/closed distributions, but what happens in practice due to rounding is another matter. :-/